sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_helpers.cs

```diff
@@ -86,11 +75,46 @@ public virtual Uri GenerateClientAccessUri(TimeSpan expiresAfter = default, stri
             {
                 expiresAt += expiresAfter;
             }
-            return GenerateClientAccessUri(expiresAt, userId, roles);
+            return GenerateClientAccessUri(expiresAt, userId, roles, cancellationToken);
+        }
+
+        private async Task<Uri> GenerateClientAccessUriInternal(
+            DateTimeOffset expiresAt,
+            string userId = default,
+            IEnumerable<string> roles = default,
+            bool async = true,
+            CancellationToken cancellationToken = default)
+        {
+            string token;
+
+            if (_tokenCredential != null)
+            {
+                RequestOptions options = new() { CancellationToken = cancellationToken };
+
+                Response clientTokenResponse = async ?
+                    await GenerateClientTokenImplAsync(userId, roles, expiresAt.Minute, options).ConfigureAwait(false) :
+                    GenerateClientTokenImpl(userId, roles, expiresAt.Minute, options);
+                token = JsonDocument.Parse(clientTokenResponse.Content).RootElement.GetProperty(ClientTokenResponseTokenPropertyName).GetString();
+            }
+            else if (_credential != null)
+            {
+                token = GenerateTokenFromAzureKeyCredential(expiresAt, userId, roles);
+            }
+            else
+            {
+                throw new InvalidOperationException($"Exactly one of {typeof(TokenCredential)} or {typeof(AzureKeyCredential)} must be set.");
```

Comment:
Maybe more like 
```suggestion
                throw new InvalidOperationException($"{nameof(WebPubSubClient} must be constructed with either a {typeof(TokenCredential)} or {typeof(AzureKeyCredential)} to generate client access URIs.");
```
since there's no property people can set?